wand bins patch: fix #479

mkborregaard opened the pull request:
Seemed generally bitrotted. Missing a dep for the FFT, the pdf function was changed in Distributions long ago. For wand bins to actually work also requires https://github.com/JuliaPlots/Plots.jl/pull/4076 in Plots (version XX)

sethaxen commented:
I'd prefer we just used the [AbstractFFTs](https://github.com/JuliaMath/AbstractFFTs.jl) interface. Once one of the packages that implements the interface is loaded, there's no way for the user to switch between them, and the whole point of the interface package is to not coerce the user to a choice. So we can document that to use wand bins, the user should load an FFT package  (e.g. FFTW or FastTransforms). And if one of the other packages in their environment already did, then great, it already works.

mkborregaard commented:
Are you sure? That sounds strange, why would the user not be able to choose a different FFT than the one we use internally here? Not really keen on hoisting the dep management manually onto the user.

sethaxen commented:
> Are you sure? That sounds strange, why would the user not be able to choose a different FFT than the one we use internally here?

Currently only if the user constructs their own FFT plan can they coerce the backend. The `fft` function takes no backend argument, so it ends up using one of the packages that has been loaded but the user can't control which. See https://github.com/JuliaMath/AbstractFFTs.jl/issues/32

> Not really keen on hoisting the dep management manually onto the user.

I think this is standard for FFT though. I'll look into it some more.

mkborregaard commented:
So the FFTW dep is a bit strange to me - this functionality obviously worked when it was added, so `fft` must have been in the namespace without FFTW at that time. What package defined that?

piever commented:
We depend on KernelDensity for the smooth density plots, and they load FFTW ([see here](https://github.com/JuliaStats/KernelDensity.jl/blob/922dfffb99d54eb1f9f88f778d1ac44eb0dcc738/src/KernelDensity.jl#L10)), so even if we get the function from AbstractFFT (which I guess we might), there will already be the FFTW backend loaded due to KernelDensity.

I personally would also prefer it if KernelDensity was FFTW backend agnostic (due to the MKL dependency of FFTW.jl), but 
1. that should be probably discussed over at KernelDensity.jl, and
2. it would IMO only really make sense once there is more than one implementation of FFT for cpu arrays.

sethaxen commented:
> So the FFTW dep is a bit strange to me - this functionality obviously worked when it was added, so `fft` must have been in the namespace without FFTW at that time. What package defined that?

According to the [AbstractFFTs docs](https://juliamath.github.io/AbstractFFTs.jl/stable/), it used to be in the stdlib for Julia 0.6 and earlier. StatsPlots dropped support for v0.6 in [v0.8.0](https://github.com/JuliaPlots/StatsPlots.jl/releases/tag/v0.8.0). Wand bins were added in v0.5.0.

sethaxen commented:
> We depend on KernelDensity for the smooth density plots, and they load FFTW ([see here](https://github.com/JuliaStats/KernelDensity.jl/blob/922dfffb99d54eb1f9f88f778d1ac44eb0dcc738/src/KernelDensity.jl#L10)), so even if we get the function from AbstractFFT (which I guess we might), there will already be the FFTW backend loaded due to KernelDensity.

Ah okay, well that then makes this discussion moot. But I would prefer to just depend on AbstractFFTs directly then.

mkborregaard commented:
@sethaxen I understand your concern as it seems we are adding a new binary dep, but actually KernelDensity here already depends directly on FFTW, so there is no difference in terms of footprint. If you prefer, I can do `import KernelDensity.FFTW: fft`? Though I think it is better practice to make the dep explicit

mkborregaard commented:
@sethaxen alright, I've updated the dep to AbstractFFTs. Seems to work perfectly. I also added a test that isn't numerically precise (because one shouldn't assume random numbers in tests) but just sees if the function runs essentially.
For some reason the Distribution tests fail on my machine when I run the StatsPlots tests locally, but that is orthogonal.

sethaxen commented on src/hist.jl:
```diff
@@ -207,12 +207,12 @@ function bkfe(gcounts, drv, bandwidth, range_x)
 
     tau = 4 + drv
-    L = min(floor(tau*h/delta), M)
+    L = min(floor(Int, tau*h/delta), M)
```
```suggestion
    L = min(Int(fld(tau*h, delta)), M)
```

mkborregaard replied:
Why is that better? I've commited your suggestion but don't really understand it.

sethaxen commented on src/hist.jl:
```diff
@@ -207,12 +207,12 @@ function bkfe(gcounts, drv, bandwidth, range_x)
     arg = lvec .* delta/h
 
-    kappam = pdf(Normal(),arg)/(h^(drv+1))
+    kappam = pdf.(Normal(),arg)/(h^(drv+1))
```
```suggestion
    kappam = pdf.(Normal(),arg) ./ h^(drv+1)
```

sethaxen commented on src/hist.jl:
```diff
@@ -227,7 +227,7 @@ function bkfe(gcounts, drv, bandwidth, range_x)
     ## Now combine weights and counts to obtain estimate
     ## we need P >= 2L+1L, M: L <= M.
-    P = 2^(ceil(log(M+L+1)/log(2)))
+    P = 2^(ceil(Int, log(M+L+1)/log(2)))
```
Looks like this is just change of base to 2:
```suggestion
    P = 2^ceil(Int, log2(M+L+1))
```

devmotion replied:
It seems this is just
```suggestion
    P = nextpow(2, M+L+1)
```

sethaxen commented on src/StatsPlots.jl:
```diff
@@ -20,7 +20,7 @@ import Clustering: Hclust, nnodes
 import MultivariateStats: MDS, eigvals, projection, principalvars,
                             principalratio, transform
-
+import AbstractFFTs: fft, ifft
```
Importing implies we overload them, which we don't.
```suggestion
using AbstractFFTs: fft, ifft
```

mkborregaard replied:
I agree, though importing anyway is the convention here.

sethaxen reviewed (changes requested):
A few minor changes

mkborregaard commented:
Changed as requested.

sethaxen reviewed (approved):
LGTM!
